src/widgets/index/__tests__/index-test.ts

```diff
@@ -810,7 +830,173 @@ See documentation: https://www.algolia.com/doc/api-reference/widgets/index/js/"
+
+      jest.spyOn(level1, 'getWidgetState');
+
+      level0.addWidgets([...widgets, level1]);
+
+      level0.init(
+        createInitOptions({
+          instantSearchInstance,
+        })
+      );
+
+      // Simulate a state change
+      level0
+        .getHelper()!
+        .setQueryParameter('query', 'Apple')
+        .setQueryParameter('page', 5);
+
+      widgets.forEach(widget => {
+        expect(widget.getWidgetState).toHaveBeenCalledTimes(2); // 2 changes
+      });
+
+      expect(level1.getWidgetState).toHaveBeenCalledTimes(0);
+    });
+
+    it('retrieves the `uiState` for the children indices', () => {
+      const level0 = index({ indexName: 'level_0_index_name' });
+      const level1 = index({ indexName: 'level_1_index_name' });
+      const level2 = index({ indexName: 'level_2_index_name' });
+      const instantSearchInstance = createInstantSearch();
+
+      level0.addWidgets([
+        createSearchBox(),
+        createPagination(),
+
+        level1.addWidgets([
+          createSearchBox(),
+          createPagination(),
+
+          level2.addWidgets([createSearchBox(), createPagination()]),
+        ]),
+      ]);
+
+      level0.init(
+        createInitOptions({
+          instantSearchInstance,
+        })
+      );
+
+      // Simulate a state change
+      level0
+        .getHelper()!
+        .setQueryParameter('query', 'Apple')
+        .setQueryParameter('page', 5);
+
+      level1
+        .getHelper()!
+        .setQueryParameter('query', 'Apple iPhone')
+        .setQueryParameter('page', 7);
+
+      level2
+        .getHelper()!
+        .setQueryParameter('query', 'Apple iPhone 5S')
+        .setQueryParameter('page', 9);
+
+      expect(level0.getWidgetState({})).toEqual({
+        // eslint-disable-next-line @typescript-eslint/camelcase
+        level_0_index_name: {
+          query: 'Apple',
+          page: 5,
+        },
+        // eslint-disable-next-line @typescript-eslint/camelcase
+        level_1_index_name: {
+          query: 'Apple iPhone',
+          page: 7,
+        },
+        // eslint-disable-next-line @typescript-eslint/camelcase
+        level_2_index_name: {
+          query: 'Apple iPhone 5S',
+          page: 9,
+        },
+      });
+
+      expect(level1.getWidgetState({})).toEqual({
+        // eslint-disable-next-line @typescript-eslint/camelcase
+        level_1_index_name: {
+          query: 'Apple iPhone',
+          page: 7,
+        },
+        // eslint-disable-next-line @typescript-eslint/camelcase
+        level_2_index_name: {
+          query: 'Apple iPhone 5S',
+          page: 9,
+        },
+      });
+
+      expect(level2.getWidgetState({})).toEqual({
+        // eslint-disable-next-line @typescript-eslint/camelcase
+        level_2_index_name: {
+          query: 'Apple iPhone 5S',
+          page: 9,
+        },
+      });
+    });
+
+    it('retrieves the `uiState` for the children indices with empty indices omit', () => {
```

Comment:
Would it make sense to keep an empty object in the UI state when the index is added to the instance but doesn't add any widgets?

IMO that's an indication that the index is mounted, but doesn't yet host any widgets. Maybe this can be leveraged? I don't have the full picture yet but maybe for SSR?

What I'm trying to say is that since we still don't _exactly_ know how we/people are going to leverage this API, perhaps removing empty keys is an enhancement that we can make once we're sure this cannot be used.

Reply:
Yes it's possible to keep it. Not sure it will be useful for SSR but at least it will be easier to perform a check at the state mapping level to detect whether we are in single/multi index context.

Reply:
https://github.com/algolia/instantsearch.js/pull/3997/commits/000bd386babc9c00d20d3b91c0f230ea8b1de73e